bciers/libs/components/src/form/MultiStepBase.tsx

```diff
@@ -54,29 +56,50 @@ const MultiStepBase = ({
   uiSchema,
   submitButtonDisabled,
   customValidate,
+  successComponent,
+  firstStepExtraHandling,
 }: MultiStepBaseProps) => {
   const [isSubmitting, setIsSubmitting] = useState(false);
+  const [isSuccessfullySubmitted, setIsSuccessfullySubmitted] = useState(false);
   const [isEditMode, setIsEditMode] = useState(false);
-
+  const [error, setError] = useState(parentError);
   const router = useRouter();
   const isNotFinalStep = step !== steps?.length;
 
   const stepIndex = step - 1;
 
-  // Set isSubmitting to true to disable submit buttons and prevent multiple form submissions
   const submitHandler = async (data: any) => {
+    // Set isSubmitting to true to disable submit buttons and prevent multiple form submissions
     setIsSubmitting(true);
-    const response = await onSubmit(data);
-
-    // If there is an error, set isSubmitting to false to re-enable submit buttons
-    // and allow user to attempt to re-submit the form
-    if (response?.error) {
+    // Clear any old errors
+    setError(undefined);
+    try {
+      const response = await onSubmit(data);
+      if (response?.error) {
+        setError(response?.error);
+        return;
+      }
+      // In some cases, for the first step, we need to do something beyond simply redirecting to the baseUrl after successful onSubmit.
+      if (step === 1 && firstStepExtraHandling) {
+        firstStepExtraHandling(response);
+        return;
+      }
+      if (isNotFinalStep && baseUrl) {
+        const nextStepUrl = `${baseUrl}/${step + 1}${
+          baseUrlParams ? `?${baseUrlParams}` : ""
+        }`;
+        router.push(nextStepUrl);
+        return;
+      }
+      // Last step
+      if (!isNotFinalStep) {
+        setIsSuccessfullySubmitted(true);
+      }
+    } catch (err) {
+      // Handle any errors beyond response.error
+      setError("An unexpected error occurred. Please try again.");
+    } finally {
```

Comment:
```suggestion
  const submitHandler = async (data: any) => {
    // Set isSubmitting to true to disable submit buttons and prevent multiple form submissions
    setIsSubmitting(true);
    // Clear any old errors
    setError(undefined);
    try {
      const response = await onSubmit(data);
      if (response?.error) {
        setError(response?.error);
        return response;
      }

      if (baseUrl) {
        const nextStepUrl = `${baseUrl}/${step + 1}${
          baseUrlParams ? `?${baseUrlParams}` : ""
        }`;
        router.push(nextStepUrl);
        return response;
      }
      return response;
    } catch (err) {
      // Handle any errors beyond response.error
      setError("An unexpected error occurred. Please try again.");
    } finally {
      setIsSubmitting(false);
    }
  };
```

For your consideration: I think this is getting overly complicated and hard to follow and will be difficult to maintain in the future. I think we should just not handle navigation if the `baseUrl` doesn't exist, and return the response to the parent component if they need that information.

Reply:
> return the response to the parent component if they need that information
What do you mean by this? I've only ever done this with state, and that's not working here (maybe because it takes too long to get the response?) Or do you mean like your suggestion above, which does work:
```
onSubmit={async (data: any) => {
        const response = await handleSubmit(data);
        const nextStepUrl = `/register-an-operation/${response.id}/${step + 1}`;
        router.push(nextStepUrl);
      }}
```

Reply:
Yeah that's what I meant, in your MultiStepBase changes we just do a plain return. In the suggestion I made we `return reponse` and can access it in the parent component as shown here.

Reply:
Sorry I think I misunderstood you, though I think we can pass `handleSubmit` and to a `.then` to get the response from in the parent?